#!/usr/bin/env Rscript

# Authors: Rebekka Mueller, David Widmann
# 2016-02-01
# v0.2

"Merge reads of samples by excluding samples of worst quality cluster in AneuFinder analysis.

Usage:
  merge_singlecells.R [options] <model>...
  merge_singlecells.R -h | --help
  merge_singlecells.R --version

Arguments:
  <model>                AneuFinder model (DNAcopy or HMM).

Options:
  -h, --help             Show this screen.
  --version              Show version.
  -f, --force            Force to overwrite the output file if present.
  -i, --input <input>    Input folder with aligned reads as BAM files [default: .].
  -o, --output <output>  Output file [default: ./merged_best.bam].
" -> doc

suppressPackageStartupMessages(library(docopt))

opt <- docopt(doc, version="0.2")

cat("Options:\n")
cat("\tmodel =", opt$model, "\n")
cat("\tforce =", opt$force, "\n")
cat("\tinput =", opt$input, "\n")
cat("\toutput =", opt$output, "\n\n")

# Set up output directory
if (!dir.exists(dirname(opt$output)))
  dir.create(dirname(opt$output), recursive=TRUE)

if (!opt$force && any(file.exists(paste0(opt$output, c('','.bai','.log')))))
  stop("Output file already exists. Specify '-f' or '--force' on the commandline to overwrite it.")

# Path of log file
log.file <- paste0(opt$output, '.log')
sink(file=log.file, split=TRUE)

# Load AneuFinder
cat("Loading AneuFinder ...\n\n")
suppressPackageStartupMessages(library(AneuFinder))
suppressPackageStartupMessages(library(mclust))

# Load AneuFinder models
models <- loadFromFiles(unique(opt$model), check.class='aneuHMM')

# Remove models without corresponding BAM file
names(models) <- sapply(models, "[", "ID")
bam.files <- list.files(path=opt$input, pattern="\\.bam$", ignore.case=TRUE)
models.missing <- setdiff(names(models), bam.files)
if (length(models.missing) > 0)
  warning(paste("Could not find the following BAM files:",
                models.missing, sep="\n"))
models <- models[names(models) %in% bam.files]

# Set quality measures depending on type of models
models.method <- sapply(models, function (x) {
                          match.call(definition = findCNVs <- function(method, ...) {},
                                     attr(x, "call"))$method})
if (any(models.method=="dnacopy")) {
  measures <- c('spikiness','num.segments','entropy','sos')
  cat("Found AneuFinder model generated by DNAcopy.\n")
} else {
  measures <- c('spikiness','num.segments','entropy','bhattacharyya','sos')
}
cat("Using quality measures ", toString(measures), ".\n\n", sep="")

# Cluster selected HMM files
cl <- clusterByQuality(models, measures=measures)
cl.length <- length(cl$classification)
cat("Found", cl.length, "clusters with the following parameters:\n")
print(cl$parameters)
cat("\n")

# Remove cluster of lowest quality
if (cl.length==1) {
  warning("Can not remove lowest scoring cluster. Only one cluster found.\n")
} else {
  removed.samples <- cl$classification[[cl.length]]
  cat("Removing the following samples contained in cluster ", cl.length, ":\n", sep="")
  cat(removed.samples, sep="\n")
  cat("\n")

  merged.samples <- sort(unlist(cl$classification[-cl.length]))
  cat("Merging the following samples contained in the other clusters:",
      merged.samples, sep="\n")
  cat("\n")
}

# Merge samples in good clusters
cat("Loading Rsamtools ...\n\n")
suppressPackageStartupMessages(library(Rsamtools))
cat("Merging samples ...\n")
mergeBam(files=file.path(opt$input, merged.samples), destination=opt$output,
         overwrite=TRUE, indexDestination=TRUE)

cat("\nDone.\n")
